### PR TITLE
feat(step-functions): number and string coerce

### PR DESCRIPTION
**Pull request by thantos**
```ts
Number("1") // 1
+"1" // 1
String(1) // "1"
```

### CODE REVIEW

File: src/asl/synth.ts [hunk 1]
```diff
@@ -1894,12 +1894,77 @@ export class ASL {
               return { value: !!argOutput.value, containsJsonPath: false };
             }
           });
+        } else if (ref === String) {
+          const [arg] = expr.args;
```
Comment (sam-goodwin) [inline src/asl/synth.ts:1898]: Do you want to check for `String(undefined)`, i.e. where `expr.args` is an `UndefinedLiteralExpr`?

File: src/asl/synth.ts [hunk 2]
```diff
@@ -1894,12 +1894,77 @@ export class ASL {
...
+            if (ASLGraph.isJsonPath(argOutput) || argOutput.containsJsonPath) {
+              const temp = this.newHeapVariable();
+              const jsonPathStates = this.normalizeOutputToJsonPath(argOutput);
+              const { jsonPath } = ASLGraph.isJsonPath(jsonPathStates)
+                ? jsonPathStates
+                : jsonPathStates.output;
+
+              return {
+                ...ASLGraph.joinSubStates(arg, jsonPathStates, {
+                  // if this was a literal, we know it was an object, not a string
```
Comment (sam-goodwin) [inline src/asl/synth.ts:1917]: why?

File: src/asl/synth.ts [hunk 3]
```diff
@@ -1894,12 +1894,77 @@ export class ASL {
...
+                    checkString: {
+                      Type: "Choice",
+                      Choices: [{ ...ASL.isString(jsonPath), Next: "assign" }],
+                      Default: "format",
+                    },
+                    format: {
+                      Type: "Pass",
+                      Parameters: {
+                        "str.$": `States.JsonToString(${jsonPath})`,
+                      },
```
Comment (sam-goodwin) [inline src/asl/synth.ts:1932]: Converting it to a JSON string?

File: src/asl/synth.ts [hunk 4]
```diff
@@ -1894,12 +1894,77 @@ export class ASL {
...
+        } else if (ref === Number) {
+          const [arg] = expr.args;
+          return this.evalToNumber(arg?.expr);
         }
       }
       throw new Error(
-        `call must be a service call or list .slice, .map, .forEach or .filter: ${toStateName(
-          expr
-        )}`
+        `call must be an integration call, list(.slice, .map, .forEach, .filter, or ` +
```
Comment (sam-goodwin) [inline src/asl/synth.ts:1964]: typeo, `list(`

File: src/asl/synth.ts [hunk 5]
```diff
@@ -2470,6 +2536,139 @@ export class ASL {
...
+                  ),
+                  Next: "zero",
+                },
+                // Number(null) => 0
+                { ...ASL.isNull(output.jsonPath), Next: "zero" },
+                // Number("1") => 1
+                { ...ASL.isString(output.jsonPath), Next: "format" },
+                // Number(1) => 1
+                { ...ASL.isNumeric(output.jsonPath), Next: "assign" },
+                // Number(true) => 1
```
Comment (sam-goodwin) [inline src/asl/synth.ts:2598]: This reads so well, great job. It's almost coding haha. We should expose it and document it. Great escape hatch DSL.

**Review by sam-goodwin — APPROVED** [review]

File: src/asl/synth.ts (hunk 2 above)
Comment (thantos) [reply src/asl/synth.ts:1917]: `containsJsonPath` is only possible for an Object literal

File: src/asl/synth.ts (hunk 3 above)
Comment (thantos) [reply src/asl/synth.ts:1932]: This worked as I expected for all of the tests... but maybe I'm missing something... I'll add more tests. I think the edge case will be that most objects just print `[object object]` instead of getting stringified.

File: src/asl/synth.ts (hunk 4 above)
Comment (thantos) [reply src/asl/synth.ts:1964]: I meant to do that, was trying to show `list(.slice, .map, ...)`

File: src/asl/synth.ts (hunk 5 above)
Comment (thantos) [reply src/asl/synth.ts:2598]: Thanks, love how easy it is.

File: src/asl/synth.ts (hunk 3 above)
Comment (thantos) [reply src/asl/synth.ts:1932]: hmm

image

File: src/asl/synth.ts (hunk 1 above)
Comment (thantos) [reply src/asl/synth.ts:1898]: Currently that will show `Step Functions does not support undefined`, but I can make it work at least.

File: src/asl/synth.ts (hunk 2 above)
Comment (thantos) [reply src/asl/synth.ts:1917]: >                   // if this was a literal, we know it is an object (only objects have containsJsonPath:true), not a string
